File: packages/agent/agent/tools/reject_batch.ts
import { getDb } from "@clara-financas/db";
import { batches, transactions } from "@clara-financas/db/schema/ledger";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { and, eq } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { notFound, refused } from "../lib/errors";
import { requireTenantCaller, tenantIdOf } from "../lib/tenant";

/**
 * Descartar uma fatura proposta.
 *
 * O botão "Rejeitar" do cartão de conferência existia na interface e não tinha
 * tool por trás: ele mandava a FRASE "Descarte o lote X. Não quero registrar
 * essa fatura." para o chat. A Clara então respondia que tinha descartado — e
 * nada acontecia no banco. O lote seguia `proposed` para sempre, reaparecendo
 * no snapshot de todo turno e na lista de faturas como pendência viva.
 *
 * O status `rejected` já existia no schema desde o começo, sem nenhum código
 * que o escrevesse.
 */
export default defineTool({
  description:
    "Requests approval to discard a proposed invoice that will not be recorded. Use when the person says the invoice is wrong, duplicated, or that they do not want it. The draft entries are removed and the batch is marked as rejected, so it stops showing up as pending.",
  inputSchema: z.object({
    batchId: z.string().min(1),
    reason: z
      .string()
      .min(1)
      .describe("Why it is being discarded, in Brazilian Portuguese. Shown on the card."),
  }),

  approval: (ctx) => {
    const current = tenantIdOf(ctx.session.auth.current);
    if (current === undefined || current !== tenantIdOf(ctx.session.auth.initiator)) {
      return { type: "denied", reason: "A sessão não está fixada a um único usuário." };
    }
    return "user-approval";
  },

  async execute(input, ctx) {
    const { tenantId, userId } = requireTenantCaller(ctx);

    return forTenant(
      tenantId,
      async (tx) => {
        const [batch] = await tx
          .select()
          .from(batches)
          .where(and(eq(batches.id, input.batchId), eq(batches.tenantId, tenantId)))
          .limit(1);

        if (batch === undefined) {
          return notFound("lote_nao_encontrado", `Nenhuma fatura com o id ${input.batchId}.`, {
            hint: "Confira os batchId do estado do razão, ou chame list_invoices.",
          });
        }
        if (batch.status === "rejected") {
          return {
            batchId: batch.id,
            status: "rejected" as const,
            alreadyRejected: true as const,
          };
        }
        if (batch.status === "confirmed") {
          return refused(
            "lote_ja_registrado",
            "Esta fatura já está no razão, e o que foi registrado não se apaga.",
            {
              hint: "Para corrigir um valor dela, chame create_adjustment. Para revisar item a item, chame read_batch.",
            },
          );
        }

        // Apagar só as linhas em rascunho — as confirmadas nem chegam aqui, mas
        // o filtro explícito é o que garante que o trigger de imutabilidade
        // nunca precise recusar nada.
        const removed = await tx
          .delete(transactions)
          .where(
            and(
              eq(transactions.batchId, batch.id),
              eq(transactions.tenantId, tenantId),
              eq(transactions.status, "proposed"),
            ),
          )
          .returning({ id: transactions.id });

        await tx
          .update(batches)
          .set({ status: "rejected", approvedBy: `human:${userId}`, approvedAt: new Date() })
          .where(and(eq(batches.id, batch.id), eq(batches.tenantId, tenantId)));

        return {
          batchId: batch.id,
          status: "rejected" as const,
          discardedTransactions: removed.length,
          reason: input.reason,
          auditedBy: `human:${userId}`,
          // O documento fica: reenviar o mesmo PDF é barrado pelo hash, e sem
          // esta nota o modelo prometeria um reenvio que a deduplicação recusa.
          note: "O documento continua registrado; para propor de novo a partir dele, use propose_batch.",
        };
      },
      getDb(),
    );
  },
});

File: services/agent/clara/tools/reject_batch.py
"""Descartar uma fatura proposta — porta de `agent/tools/reject_batch.ts`.

O status `rejected` existe no schema desde o começo. Esta é a única função
que o escreve. Sem ela, "descartei" dito em prosa não muda nada no banco: o
rascunho segue `proposed` para sempre e reaparece no snapshot de todo turno.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime

from sqlalchemy import delete, select, update
from sqlalchemy.orm import Session

from clara.db.models import Batch, Transaction
from clara.tools.errors import ToolError, not_found, refused


@dataclass
class RejectReceipt:
    batch_id: str
    status: str
    already_rejected: bool = False
    discarded_transactions: int = 0
    reason: str | None = None
    audited_by: str | None = None
    note: str | None = None


def reject_batch(
    session: Session, tenant_id: str, user_id: str, batch_id: str, reason: str
) -> RejectReceipt | ToolError:
    batch = session.execute(
        select(Batch).where(Batch.id == batch_id, Batch.tenant_id == tenant_id)
    ).scalar_one_or_none()

    if batch is None:
        return not_found(
            "lote_nao_encontrado", f"Nenhuma fatura com o id {batch_id}.",
            hint="Confira os batch_id do estado do razão, ou chame list_invoices.",
        )

    if batch.status == "rejected":
        return RejectReceipt(batch_id=batch.id, status="rejected", already_rejected=True)

    if batch.status == "confirmed":
        return refused(
            "lote_ja_registrado",
            "Esta fatura já está no razão, e o que foi registrado não se apaga.",
            hint=(
                "Para corrigir um valor dela, chame create_adjustment. "
                "Para revisar item a item, chame read_batch."
            ),
        )

    # Apagar só as linhas em rascunho: o filtro explícito garante que o
    # trigger de imutabilidade nunca precise recusar nada aqui.
    removed = session.execute(
        delete(Transaction)
        .where(
            Transaction.batch_id == batch.id,
            Transaction.tenant_id == tenant_id,
            Transaction.status == "proposed",
        )
        .returning(Transaction.id)
    ).scalars().all()

    session.execute(
        update(Batch)
        .where(Batch.id == batch.id, Batch.tenant_id == tenant_id)
        .values(status="rejected", approved_by=f"human:{user_id}", approved_at=datetime.now(UTC))
    )
    session.flush()

    return RejectReceipt(
        batch_id=batch.id, status="rejected", discarded_transactions=len(removed),
        reason=reason, audited_by=f"human:{user_id}",
        note=(
            "O documento continua registrado; "
            "para propor de novo a partir dele, use propose_batch."
        ),
    )
